File: core/aggregators/mactechnews/aggregator.py
"""MacTechNews aggregator implementation."""

import re
from typing import Any, Dict, List
from urllib.parse import urljoin

from bs4 import BeautifulSoup

from ..utils import (
    clean_html,
    format_article_content,
    remove_image_by_url,
    sanitize_class_names,
)
from ..utils.youtube import proxy_youtube_embeds
from ..website import FullWebsiteAggregator
from .comment_extractor import extract_comments
from .multipage_handler import detect_pagination, fetch_all_pages

# Recurring "TechTicker:" link-roundup posts are noise. Prefix match, case
# sensitive -- mirroring iOS's shouldInclude. A looser `contains` would drop
# legitimate articles that merely mention the word. Heise's title skip-list is
# deliberately the opposite (substring, case-insensitive); do not unify them.
TECHTICKER_TITLE_PREFIX = "TechTicker:"


class MactechnewsAggregator(FullWebsiteAggregator):
    """Aggregator for MacTechNews (mactechnews.de)."""

    brand_site_url = "https://www.mactechnews.de/"

    def __init__(self, feed):
        super().__init__(feed)
        if not self.identifier or self.identifier == "":
            self.identifier = "https://www.mactechnews.de/Rss/News.x"

    def get_source_url(self) -> str:
        return "https://www.mactechnews.de"

    @classmethod
    def resolves_feed_url(cls) -> bool:
        """Fixed-brand scraper: the identifier is its own hardcoded feed URL."""
        return False

    @classmethod
    def get_configuration_fields(cls) -> Dict[str, Any]:
        """Configuration options for MacTechNews aggregator."""
        from django import forms

        return {
            "combine_pages": forms.BooleanField(
                initial=True,
                label="Combine Multi-page Articles",
                help_text="Automatically fetch and combine all pages of a multi-page article.",
                required=False,
            ),
            "include_comments": forms.BooleanField(
                initial=True,
                label="Include Comments",
                help_text="Extract user comments from article pages.",
                required=False,
            ),
            "max_comments": forms.IntegerField(
                initial=5,
                label="Max Comments",
                help_text="Maximum number of comments to extract per article.",
                required=False,
                min_value=0,
                max_value=20,
            ),
        }

    def filter_articles(self, articles: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Drop TechTicker link roundups, then apply the base age filter."""
        articles = super().filter_articles(articles)

        kept = []
        for article in articles:
            if article.get("name", "").startswith(TECHTICKER_TITLE_PREFIX):
                self.logger.info(
                    f"[filter_articles] Skipping TechTicker roundup: {article.get('name')}"
                )
                continue
            kept.append(article)

        return kept

    # Multipage articles are combined by fetch_all_pages into sibling
    # .MtnArticle containers (one per page), so every match must be kept --
    # keeping only the first would silently truncate to page 1 after fetching
    # every page over the network. A single-page article has exactly one
    # match, so the union degenerates to the same result.
    uses_first_content_match = False

    # Main content container
    content_selectors = [".MtnArticle"]

    # Selectors to strip
    selectors_to_remove = [
        ".NewsPictureMobile",
        "aside",
        "script",
        "style",
        "iframe",
        "noscript",
        "svg",
        "header",  # Remove article header (title, meta) to avoid duplication
        ".TexticonBox.Right",  # Remove sidebars/summary boxes inside content
    ]

    @staticmethod
    def _extract_mtn_image_id(url: str) -> str | None:
        """Extract numeric image ID from mactechnews image URLs.

        URLs follow the pattern: Name.{numeric_id}.{ext}
        e.g. Cover-Raumakustik.592736.jpg -> 592736
             Bild.592736.jpg -> 592736
        """
        match = re.search(r"\.(\d{5,})\.\w+$", url)
        return match.group(1) if match else None

    def fetch_article_content(self, url: str) -> str:
        """Fetch article content, handling multi-page articles."""
        self.logger.debug(f"[fetch_article_content] Starting for URL: {url}")

        combine_pages = self.feed.options.get("combine_pages", True)

        # Fetch first page
        first_page_html = super().fetch_article_content(url)
        self.logger.debug(
            f"[fetch_article_content] First page fetched ({len(first_page_html)} bytes)"
        )

        if not combine_pages:
            self.logger.info(f"[fetch_article_content] Multi-page combination disabled for {url}")
            return first_page_html

        # Detect pagination
        page_numbers = detect_pagination(first_page_html, self.logger)

        if len(page_numbers) <= 1:
            self.logger.info(f"[fetch_article_content] Single page article for {url}")
            return first_page_html

        # Multi-page article - fetch and combine all pages
        self.logger.info(f"[fetch_article_content] Multi-page article: {len(page_numbers)} pages")

        combined_html = fetch_all_pages(
            base_url=url,
            page_numbers=page_numbers,
            content_selectors=self.content_selectors,
            fetcher=lambda page_url: super(MactechnewsAggregator, self).fetch_article_content(
                page_url
            ),
            logger=self.logger,
            first_page_html=first_page_html,
        )

        self.logger.debug(f"[fetch_article_content] Combined HTML: {len(combined_html)} bytes")
        return combined_html

    def process_content(self, html: str, article: Dict[str, Any]) -> str:
        """Process content with relative URL resolution, header dedup, and comments."""
        soup = BeautifulSoup(html, "html.parser")
        base_url = article["identifier"]

        # Remove content images that duplicate the header image.
        # mactechnews uses the same numeric ID across different image variants
        header_data = article.get("header_data")
        if header_data and header_data.image_url:
            header_image_id = self._extract_mtn_image_id(header_data.image_url)
            if header_image_id:
                for img in soup.find_all("img"):
                    src = img.get("src")
                    if (
                        src
                        and not isinstance(src, list)
                        and self._extract_mtn_image_id(src) == header_image_id
                    ):
                        img.decompose()

        # Resolve relative URLs for images
        for img in soup.find_all("img"):
            src = img.get("src")
            if (
                src
                and not isinstance(src, list)
                and not src.startswith(("http://", "https://", "data:"))
            ):
                img["src"] = urljoin(base_url, str(src))

        # Resolve relative URLs for links
        for a in soup.find_all("a"):
            href = a.get("href")
            if (
                href
                and not isinstance(href, list)
                and not href.startswith(("http://", "https://", "mailto:", "tel:", "#"))
            ):
                a["href"] = urljoin(base_url, str(href))

        # Proxy YouTube embeds
        proxy_youtube_embeds(soup)

        # Remove header image from content if extracted
        if header_data and header_data.image_url:
            remove_image_by_url(soup, header_data.image_url)

        # Sanitize class names
        sanitize_class_names(soup)

        # Clean HTML
        cleaned = clean_html(str(soup))

        # Determine header image URL
        header_image_url = header_data.image_ref if header_data else None

        # Extract comments from the raw (full) HTML
        comments_html = None
        include_comments = self.feed.options.get("include_comments", True)
        max_comments = self.feed.options.get("max_comments", 5)

        if include_comments:
            try:
                raw_html = article.get("raw_content", "")
                if raw_html:
                    comments_html = extract_comments(
                        raw_html,
                        article["identifier"],
                        max_comments=max_comments,
                        logger=self.logger,
                    )
            except Exception as e:
                self.logger.warning(f"[process_content] Failed to extract comments: {e}")

        # Format with header, content, comments, and footer
        formatted = format_article_content(
            cleaned,
            title=article["name"],
            url=article["identifier"],
            header_image_url=header_image_url,
            comments_content=comments_html,
        )

        return formatted

File: src/lib/aggregators/sites/mactechnews/aggregator.ts
import * as cheerio from "cheerio";
import { FeedLike, RawArticle } from "../../base";
import { cleanHtml, removeImageByUrl, sanitizeClassNames } from "../../extract/clean";
import { formatArticleContent } from "../../extract/format";
import { getHeaderImageRef } from "../../header/context";
import { FirstPageStash, fetchAllPages } from "../../multipage";
import { FullWebsiteAggregator, proxyYoutubeEmbeds } from "../../website";
import { extractComments } from "./comments";
import { buildPageUrl, detectPagination } from "./multipage";

export const TECHTICKER_TITLE_PREFIX = "TechTicker:";

/**
 * Extract numeric image ID from mactechnews image URLs.
 *
 * URLs follow the pattern: Name.{numeric_id}.{ext}
 * e.g. Cover-Raumakustik.592736.jpg -> 592736
 *      Bild.592736.jpg -> 592736
 */
export function extractMtnImageId(url: string): string | null {
  const match = url.match(/\.(\d{5,})\.\w+$/);
  return match ? match[1] : null;
}

export class MactechnewsAggregator extends FullWebsiteAggregator {
  static brandSiteUrl = "https://www.mactechnews.de/";

  static getDefaultIdentifier(): string {
    return "https://www.mactechnews.de/Rss/News.x";
  }

  static getIdentifierChoices(): Array<[string, string]> {
    return [
      ["https://www.mactechnews.de/Rss/News.x", "News"],
      ["https://www.mactechnews.de/Rss/Rewind.x", "Rewind"],
      ["https://www.mactechnews.de/Rss/Journals.x", "Journals"],
    ];
  }

  static resolvesFeedUrl(): boolean {
    return false;
  }

  static getConfigurationFields(): Record<string, unknown> {
    return {
      combine_pages: {
        type: "boolean",
        initial: true,
        label: "Combine Multi-page Articles",
        help_text: "Automatically fetch and combine all pages of a multi-page article.",
        required: false,
      },
      include_comments: {
        type: "boolean",
        initial: true,
        label: "Include Comments",
        help_text: "Extract user comments from article pages.",
        required: false,
      },
      max_comments: {
        type: "number",
        initial: 5,
        label: "Max Comments",
        help_text: "Maximum number of comments to extract per article.",
        required: false,
        min_value: 0,
        max_value: 20,
      },
    };
  }

  static contentSelectors = [".MtnArticle"];
  protected contentSelectors = [...MactechnewsAggregator.contentSelectors];

  static selectorsToRemove = [
    ".NewsPictureMobile",
    "aside",
    "script",
    "style",
    "iframe",
    "noscript",
    "svg",
    "header",
    ".TexticonBox.Right",
  ];
  protected selectorsToRemove = [...MactechnewsAggregator.selectorsToRemove];

  usesFirstContentMatch = false;

  // Keyed by article URL, not a single field -- see `FirstPageStash`'s doc
  // comment in `../../multipage` for why: `enrichArticles()` runs
  // `fetchArticleContent()` for up to `this.concurrency` articles
  // concurrently on one aggregator instance.
  private firstPages = new FirstPageStash();

  constructor(feed: FeedLike) {
    super(feed);
    if (!this.identifier) {
      this.identifier = "https://www.mactechnews.de/Rss/News.x";
    }
  }

  override getSourceUrl(): string {
    return "https://www.mactechnews.de";
  }

  /**
   * The article headline. Called once per page fetch (this aggregator
   * paginates -- see `fetchArticleContent()` below); `noteSourceTitle()`'s
   * "sticky" rule (see ../../base) keeps an earlier page's match from being
   * blanked out by a later page that doesn't repeat the heading.
   */
  protected override sourceTitleFrom($: cheerio.CheerioAPI): string | null {
    const heading = $(".MtnArticle h1, .MtnArticle h2").first().text().trim();
    return heading || null;
  }

  override async filterArticles(
    articles: RawArticle[],
    clock: () => Date = () => new Date(),
  ): Promise<RawArticle[]> {
    const filtered = await super.filterArticles(articles, clock);
    const kept: RawArticle[] = [];
    for (const article of filtered) {
      if ((article.name || "").startsWith(TECHTICKER_TITLE_PREFIX)) {
        continue;
      }
      kept.push(article);
    }
    return kept;
  }

  override async fetchArticleContent(url: string): Promise<string> {
    const options = (this.feed.options as Record<string, unknown> | null) || {};
    const combinePages = options.combine_pages !== false;

    const firstPageHtml = await super.fetchArticleContent(url);
    this.firstPages.set(url, firstPageHtml);

    if (!combinePages) {
      return firstPageHtml;
    }

    const pageNumbers = detectPagination(firstPageHtml);
    if (pageNumbers.size <= 1) {
      return firstPageHtml;
    }

    const { combined } = await fetchAllPages(
      url,
      pageNumbers,
      this.getContentSelectors(),
      (pageUrl) => super.fetchArticleContent(pageUrl),
      firstPageHtml,
      buildPageUrl,
    );

    return combined;
  }

  override async processContent(html: string, article: RawArticle): Promise<string> {
    const labels = await this.chromeLabels();
    const $ = cheerio.load(html);
    const baseUrl = article.identifier;

    // Remove content images that duplicate the header image
    const headerData = article.header_data;
    if (headerData?.imageUrl) {
      const headerImageId = extractMtnImageId(headerData.imageUrl);
      if (headerImageId) {
        $("img").each((_, img) => {
          const src = $(img).attr("src");
          if (src && extractMtnImageId(src) === headerImageId) {
            $(img).remove();
          }
        });
      }
    }

    // Resolve relative URLs for images
    $("img").each((_, img) => {
      const $img = $(img);
      const src = $img.attr("src");
      if (
        src &&
        !src.startsWith("http://") &&
        !src.startsWith("https://") &&
        !src.startsWith("data:")
      ) {
        try {
          $img.attr("src", new URL(src, baseUrl).toString());
        } catch {
          // ignore
        }
      }
    });

    // Resolve relative URLs for links
    $("a").each((_, a) => {
      const $a = $(a);
      const href = $a.attr("href");
      if (
        href &&
        !href.startsWith("http://") &&
        !href.startsWith("https://") &&
        !href.startsWith("mailto:") &&
        !href.startsWith("tel:") &&
        !href.startsWith("#")
      ) {
        try {
          $a.attr("href", new URL(href, baseUrl).toString());
        } catch {
          // ignore
        }
      }
    });

    // Replace YouTube iframes with click-through facades
    await proxyYoutubeEmbeds($, labels);

    // Remove header image from content if extracted
    if (headerData?.imageUrl) {
      removeImageByUrl($, headerData.imageUrl);
    }

    // Sanitize class names
    sanitizeClassNames($);

    // Clean HTML
    const cleaned = cleanHtml($.html());

    // Determine header image URL
    const headerImageUrl = headerData ? getHeaderImageRef(headerData) : null;

    // Retrieve and clear this article's entry unconditionally -- fetchArticleContent()
    // always records one regardless of the include_comments option, so leaving the read
    // gated behind that option would leak an entry per article on every run with
    // comments disabled.
    const firstPageHtml = this.firstPages.take(article.identifier);

    // Extract comments from the raw, un-truncated first page -- never
    // `article.raw_content`. On a multi-page article that field holds
    // `fetchAllPages()`'s `combined` result (see ../../multipage.ts), which
    // only ever contains `.MtnArticle` matches from every page and never
    // `div.MtnCommentScroll`, a sibling element outside every page's content
    // selector -- reading it here lost every multi-page article's comments.
    // Falls back to `raw_content` when nothing was stashed (combine_pages
    // off, so there was only ever one page and `raw_content` already holds
    // it in full).
    let commentsHtml: string | null = null;
    const options = (this.feed.options as Record<string, unknown> | null) || {};
    const includeComments = options.include_comments !== false;
    const maxComments = typeof options.max_comments === "number" ? options.max_comments : 5;

    if (includeComments) {
      const commentSource = firstPageHtml || article.raw_content || "";
      if (commentSource) {
        // extractComments() never throws -- a selector-extraction failure is
        // caught and logged by the shared buildCommentsSection() it delegates
        // to (see src/lib/aggregators/comments/section.ts), rather than being
        // swallowed silently by a try/catch here.
        commentsHtml = extractComments(
          commentSource,
          article.identifier,
          maxComments,
          labels,
          this.onLog,
        );
      }
    }

    return formatArticleContent(
      cleaned,
      article.name,
      article.identifier,
      labels,
      headerImageUrl,
      null,
      commentsHtml,
    );
  }
}
